src/odemis/driver/actuator.py

```diff
@@ -2179,3 +2182,601 @@ def terminate(self):
+        self._executeMoves(ordered_moves)
+
+    def _doMoveRel(self, future, shift):
+        """
+        Move to the requested relative position and adjust focus if necessary
+        """
+        ordered_moves = self._getOrderedMoves(future, shift, rel=True)
+        self._executeMoves(ordered_moves)
+
+    def _getOrderedMoves(self, future, vector_value, rel=False):
+        """
+        Get the ordered movement sequence of the underlying stage and the focus to ensure no collision
+        :param vector_value: The target vector value (absolute position or shift)
+        :param rel: Whether it's relative movement or absolute
+        :return: A list of either one to two functions and their arguments ordered by execution priority
+        """
+        ordered_moves = []
+        stage_fn = self._doMoveRelStage if rel else self._doMoveAbsStage
+
+        # Determine movement and direction of the Z and Rx axes
+        z_move = self._checkZDirection(vector_value, rel)
+        rx_move = self._checkRxMovement(vector_value, rel)
+
+        # Order the stage and focus movements to ensure no collision
+        # The initial movement would increase the distance between the two stages,
+        # while the subsequent move would bring back the focus value
+        if z_move == MoveUp:
+            # Move stage first then focus
+            ordered_moves.append((stage_fn, future, vector_value))
+            if rx_move == NoRxMove:
+                # Pass None in case of absolute move (so that the focus would take it from the updated stage position)
+                target_stagez = vector_value['z'] if rel else None
+                ordered_moves.append((self._focus.adjustFocus, future, target_stagez, rel))
+            else:
+                logging.warning("Focus movement is dropped due to movement in rx != 0")
+        elif z_move == MoveDown:
+            # Move focus first then stage
+            if rx_move == NoRxMove:
+                ordered_moves.append((self._focus.adjustFocus, future, vector_value['z'], rel))
+            else:
+                logging.warning("Focus movement is dropped due to movement in rx != 0")
+            ordered_moves.append((stage_fn, future, vector_value))
+        else:
+            # No Z move => No movement in focus
+            ordered_moves.append((stage_fn, future, vector_value))
+        return ordered_moves
+
+    def _executeMoves(self, ordered_moves):
+        """
+        Call the list of move functions in respect to their ordered execution
+        :param ordered_moves: List of ordered moves functions
+        """
+        for i, ordered_move in enumerate(ordered_moves):
+            try:
+                move_func, *args = ordered_move
+                move_func(*args)
+                if i == 1:
+                    # Re-update the focus position after the second move
+                    self._focus._updatePosition(self._lensz.position.value)
```

Comment:
Sorry I wasn't clear about this. Basically there are two cases:
1. When the focus is the initial movement, it would need re-updating after the stage is subsequently moved. 
2. The second case is when the stage is moved first. If the focus is updated immediately, and the focus move tried to get the lens position from `_getLensZValue` (which depends on focus position), it returns the same lens position where it is now so it doesn't move. That's why it should wait on updating the focus until it actually moves, so updating the focus anytime the parent stage moves doesn't work. 

I moved it after the loop as you suggested.